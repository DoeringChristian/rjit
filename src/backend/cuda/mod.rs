pub mod codegen;
pub mod compress;
pub mod cuda;
pub mod cuda_core;
pub mod params;

pub use cuda::*;

#[cfg(test)]
mod test;
