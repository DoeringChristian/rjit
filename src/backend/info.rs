pub use super::*;

#[derive(Debug, Hash)]
pub struct ModuleInfo {
    pub asm: String,
    pub entry_point: String,
}
#[derive(Debug, Hash)]
pub struct HitGroupInfo {
    pub closest_hit: ModuleInfo,
    pub any_hit: Option<ModuleInfo>,
    pub intersection: Option<ModuleInfo>,
}
#[derive(Debug, Hash)]
pub struct MissGroupInfo {
    pub miss: ModuleInfo,
}
#[derive(Debug, Hash)]
pub struct SBTInfo {
    pub hit_groups: Vec<HitGroupInfo>,
    pub miss_groups: Vec<MissGroupInfo>,
}

impl<'a> From<SBTDesc<'a>> for SBTInfo {
    fn from(value: SBTDesc<'a>) -> Self {
        Self {
            hit_groups: value.hit_groups.iter().map(|g| (g).into()).collect(),
            miss_groups: value.miss_groups.iter().map(|g| (g).into()).collect(),
        }
    }
}

impl<'a> From<&'a ModuleDesc<'a>> for ModuleInfo {
    fn from(value: &'a ModuleDesc) -> Self {
        Self {
            asm: value.asm.into(),
            entry_point: value.entry_point.into(),
        }
    }
}

impl<'a> From<&'a HitGroupDesc<'a>> for HitGroupInfo {
    fn from(value: &'a HitGroupDesc<'a>) -> Self {
        Self {
            closest_hit: (&value.closest_hit).into(),
            any_hit: value.any_hit.as_ref().map(|m| m.into()),
            intersection: value.intersection.as_ref().map(|m| m.into()),
        }
    }
}
impl<'a> From<&'a MissGroupDesc<'a>> for MissGroupInfo {
    fn from(value: &'a MissGroupDesc<'a>) -> Self {
        Self {
            miss: (&value.miss).into(),
        }
    }
}
