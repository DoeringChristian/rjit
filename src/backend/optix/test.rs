use std::time::Instant;

use crate::backend::CompileOptions;
use crate::jit::Jit;
use crate::trace::{ReduceOp, Trace, VarType};
use crate::{backend, AccelDesc, GeometryDesc, HitGroupDesc, InstanceDesc, MissGroupDesc};
use anyhow::Result;

#[test]
fn refcounting() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.array(&[1f32; 10])?;
    assert_eq!(x.var().rc, 1, "rc of x should be 1 (in x)");
    let y = x.add(&x)?;
    // let y = ir::add(&x, &x);

    assert_eq!(
        x.var().rc,
        3,
        "rc of x should be 3 (2 refs in y and 1 in x)"
    );

    assert_eq!(y.var().rc, 1, "rc of y should be 1 (in y)");

    ir.schedule(&[&y]);

    assert_eq!(
        y.var().rc,
        2,
        "rc of y should be 2 (1 in y and 1 in schedule)"
    );

    ir.eval()?;

    assert_eq!(
        x.var().rc,
        1,
        "rc of x should be 1 (dependencies of y shuld be cleaned)"
    );
    assert_eq!(
        y.var().rc,
        1,
        "rc of y should be 2 (y from schedule should be cleaned)"
    );

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<f32>().unwrap(), vec![2f32; 10]);
    Ok(())
}
#[test]
fn load_add_f32() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.array(&[1f32; 10])?;
    // let y = ir::add(&x, &x);
    let y = x.add(&x)?;

    ir.schedule(&[&y]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<f32>().unwrap(), vec![2f32; 10]);
    Ok(())
}
#[test]
fn load_gather_f32() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.array(&[1f32, 2., 3., 4., 5.])?;
    let i = ir.array(&[0u32, 1, 4])?;
    let y = x.gather(&i, None)?;

    ir.schedule(&[&y]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<f32>().unwrap(), vec![1., 2., 5.]);
    Ok(())
}
#[test]
fn reindex() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.index(10);

    let i = ir.index(3);
    let c = ir.literal(2u32)?;
    let i = i.add(&c)?;

    let y = x.gather(&i, None)?;

    ir.schedule(&[&y]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<u32>().unwrap(), vec![2, 3, 4]);
    Ok(())
}
#[test]
fn index() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let i = ir.index(10);

    ir.schedule(&[&i]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(
        i.to_host::<u32>().unwrap(),
        vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    );
    Ok(())
}
#[test]
fn gather_eval() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let r = {
        let x = ir.index(3);
        let y = ir.array(&[1u32, 2, 3])?;

        // let z = ir::add(&x, &y);
        let z = x.add(&y)?;

        let r = z.gather(&ir.index(3), None)?;
        r
    };
    // schedule

    ir.schedule(&[&r]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(r.to_host::<u32>().unwrap(), vec![1, 3, 5]);
    Ok(())
}
#[test]
fn paralell() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.index(10);

    let y = ir.index(3);

    ir.schedule(&[&x, &y]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(
        x.to_host::<u32>().unwrap(),
        vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    );
    assert_eq!(y.to_host::<u32>().unwrap(), vec![0, 1, 2]);
    Ok(())
}
#[test]
fn load_gather() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.array(&[1.0f32, 2., 3.])?;

    ir.schedule(&[&x]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<f32>().unwrap(), vec![1., 2., 3.]);
    Ok(())
}
#[test]
fn eval_scatter() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.array(&[0u32, 0, 0, 0])?;
    let c = ir.literal(1u32)?;
    let x = x.add(&c)?; // x: [1, 1, 1, 1]

    let i = ir.index(3);
    let c = ir.literal(1u32)?;
    let i = i.add(&c)?; // i: [1, 2, 3]

    let y = ir.literal(2u32)?;

    y.scatter(&x, &i, None)?; // x: [1, 2, 2, 2]
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<u32>().unwrap(), vec![1, 2, 2, 2]);
    Ok(())
}
#[test]
fn scatter_twice() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.array(&[0u32, 0, 0, 0])?;

    let i = ir.index(3);
    let c = ir.literal(1u32)?;
    let i = i.add(&c)?;

    let y = ir.literal(2u32)?;

    y.scatter(&x, &i, None)?;

    let i = ir.index(2);

    let y = ir.literal(3u32)?;

    y.scatter(&x, &i, None)?;
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<u32>().unwrap(), vec![3, 3, 2, 2]);
    Ok(())
}
#[test]
fn scatter_twice_add() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.array(&[0u32, 0, 0, 0])?;

    let i = ir.index(3);
    let c = ir.literal(1u32)?;
    let i = i.add(&c)?;

    let y = ir.literal(2u32)?;

    y.scatter(&x, &i, None)?;

    let i = ir.index(2);

    let y = ir.literal(3u32)?;

    y.scatter(&x, &i, None)?;

    let c = ir.literal(1u32)?;
    let x = x.add(&c)?;

    ir.schedule(&[&x]);
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<u32>().unwrap(), vec![4, 4, 3, 3]);
    Ok(())
}
#[test]
fn scatter_reduce() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.array(&[0u32, 0, 0, 0])?;

    let i = ir.array(&[0u32, 0, 0])?;

    let y = ir.literal(1u32)?;

    y.scatter_reduce(&x, &i, None, ReduceOp::Add)?;

    ir.eval()?;

    // dbg!(&ir);

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<u32>().unwrap(), vec![3, 0, 0, 0]);
    Ok(())
}
#[test]
fn tex_lookup() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.array(&[0.5f32])?;
    let y = ir.array(&[0.5f32])?;

    let data = ir.array(&[1.0f32; 400])?;

    let tex = data.to_texture(&[10, 10], 4)?;

    let res = tex.tex_lookup(&[&x, &y])?;

    let r = res[0].clone();

    r.schedule();

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(r.to_host::<f32>().unwrap(), vec![1.]);
    Ok(())
}
#[test]
fn trace_ray() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let miss_and_closesthit_ptx = r##"
.version 8.0
.target sm_86
.address_size 64

.entry __miss__ms() {
	.reg .b32 %r<6>;
	mov.b32 %r0, 0;
	mov.b32 %r1, 0;
        
	call _optix_set_payload, (%r0, %r1);
	ret;
}

.entry __closesthit__ch() {
	.reg .b32 %i<5>;
	.reg .b32 %v<5>;
	mov.b32 %i0, 0;
	mov.b32 %i1, 1;
	mov.b32 %i2, 2;
	mov.b32 %i3, 3;
	mov.b32 %i4, 4;

        mov.b32 %v0, 1;
	call _optix_set_payload, (%i0, %v0);
        
        call (%v1), _optix_read_primitive_idx, ();
	call _optix_set_payload, (%i1, %v1);
        
        call (%v2), _optix_read_instance_id, ();
	call _optix_set_payload, (%i2, %v2);
        
	.reg .f32 %f<2>;
        call (%f0, %f1), _optix_get_triangle_barycentrics, ();
        mov.b32 %v3, %f0;
        mov.b32 %v4, %f1;
	call _optix_set_payload, (%i3, %v3);
	call _optix_set_payload, (%i4, %v4);
        
	ret;
}
"##;
    let indices = ir.array(&[0u32, 1, 2])?;
    let vertices = ir.array(&[1.0f32, 0., 1., 0., 1., 1., 1., 1., 1.])?;

    let desc = AccelDesc {
        sbt: crate::SBTDesc {
            hit_groups: &[HitGroupDesc {
                closest_hit: crate::ModuleDesc {
                    asm: miss_and_closesthit_ptx,
                    entry_point: "__closesthit__ch",
                },
                ..Default::default()
            }],
            miss_groups: &[MissGroupDesc {
                miss: crate::ModuleDesc {
                    asm: miss_and_closesthit_ptx,
                    entry_point: "__miss__ms",
                },
            }],
        },
        geometries: &[GeometryDesc::Triangles {
            vertices: &vertices,
            indices: &indices,
        }],
        instances: &[InstanceDesc {
            hit_group: 0,
            geometry: 0,
            transform: [1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0.],
        }],
    };
    let accel = ir.accel(desc)?;

    let payload = accel.trace_ray(
        &[
            &ir.literal(0u32)?,
            &ir.literal(0u32)?,
            &ir.literal(0u32)?,
            &ir.literal(0u32)?,
            &ir.literal(0u32)?,
        ],
        [
            &ir.array(&[0.6f32, 0.6])?,
            &ir.literal(0.6f32)?,
            &ir.literal(0.0f32)?,
        ],
        [
            &ir.literal(0.0f32)?,
            &ir.literal(0.0f32)?,
            &ir.array(&[1.0f32, -1.])?,
        ],
        &ir.literal(0.001f32)?,
        &ir.literal(1000.0f32)?,
        &ir.literal(0.0f32)?,
        None,
        None,
        None,
        None,
        None,
        None,
    )?;

    // for p in payload.iter() {
    //     p.schedule();
    // }
    let valid = payload[0].cast(&VarType::Bool)?;

    let u = payload[3].bitcast(&VarType::F32)?;
    let v = payload[4].bitcast(&VarType::F32)?;

    let dst = ir.array(&[0f32, 1f32])?;
    u.scatter(&dst, &ir.index(2), None)?;

    valid.schedule();
    v.schedule();
    u.schedule();

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(valid.to_host::<bool>().unwrap(), vec![true, false]);
    let u = dst.to_host::<f32>().unwrap();
    let v = v.to_host::<f32>().unwrap();
    approx::assert_ulps_eq!(u[0], 0.39999998);
    approx::assert_ulps_eq!(u[1], 0.);
    approx::assert_ulps_eq!(v[0], 0.20000005);
    approx::assert_ulps_eq!(v[1], 0.);
    Ok(())
}
#[test]
fn trace_ray_scatter() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let miss_and_closesthit_ptx = r##"
.version 8.0
.target sm_86
.address_size 64

.entry __miss__ms() {
	.reg .b32 %r<6>;
	mov.b32 %r0, 0;
	mov.b32 %r1, 0;
        
	call _optix_set_payload, (%r0, %r1);
	ret;
}

.entry __closesthit__ch() {
	.reg .b32 %i<5>;
	.reg .b32 %v<5>;
	mov.b32 %i0, 0;
	mov.b32 %i1, 1;
	mov.b32 %i2, 2;
	mov.b32 %i3, 3;
	mov.b32 %i4, 4;

        mov.b32 %v0, 1;
	call _optix_set_payload, (%i0, %v0);
        
        call (%v1), _optix_read_primitive_idx, ();
	call _optix_set_payload, (%i1, %v1);
        
        call (%v2), _optix_read_instance_id, ();
	call _optix_set_payload, (%i2, %v2);
        
	.reg .f32 %f<2>;
        call (%f0, %f1), _optix_get_triangle_barycentrics, ();
        mov.b32 %v3, %f0;
        mov.b32 %v4, %f1;
	call _optix_set_payload, (%i3, %v3);
	call _optix_set_payload, (%i4, %v4);
        
	ret;
}
"##;

    let indices = ir.array(&[0u32, 1, 2])?;
    let vertices = ir.array(&[1.0f32, 0., 1., 0., 1., 1., 1., 1., 1.])?;

    let desc = AccelDesc {
        sbt: crate::SBTDesc {
            hit_groups: &[HitGroupDesc {
                closest_hit: crate::ModuleDesc {
                    asm: miss_and_closesthit_ptx,
                    entry_point: "__closesthit__ch",
                },
                ..Default::default()
            }],
            miss_groups: &[MissGroupDesc {
                miss: crate::ModuleDesc {
                    asm: miss_and_closesthit_ptx,
                    entry_point: "__miss__ms",
                },
            }],
        },
        geometries: &[GeometryDesc::Triangles {
            vertices: &vertices,
            indices: &indices,
        }],
        instances: &[InstanceDesc {
            hit_group: 0,
            geometry: 0,
            transform: [1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0.],
        }],
    };
    let accel = ir.accel(desc)?;

    let payload = accel.trace_ray(
        &[
            &ir.literal(0u32)?,
            &ir.literal(0u32)?,
            &ir.literal(0u32)?,
            &ir.literal(0u32)?,
            &ir.literal(0u32)?,
        ],
        [
            &ir.array(&[0.6f32, 0.6])?,
            &ir.literal(0.6f32)?,
            &ir.literal(0.0f32)?,
        ],
        [
            &ir.literal(0.0f32)?,
            &ir.literal(0.0f32)?,
            &ir.array(&[1.0f32, -1.])?,
        ],
        &ir.literal(0.001f32)?,
        &ir.literal(1000.0f32)?,
        &ir.literal(0.0f32)?,
        None,
        None,
        None,
        None,
        None,
        None,
    )?;

    let valid = payload[0].cast(&VarType::Bool)?;

    let u = payload[3].bitcast(&VarType::F32)?;

    let dst = ir.array(&[0f32, 0f32])?;
    u.scatter(&dst, &ir.index(2), Some(&valid))?;

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    // assert_eq!(valid.to_host_bool(), vec![true, false]);
    let u = dst.to_host::<f32>().unwrap();
    approx::assert_ulps_eq!(u[0], 0.39999998);
    approx::assert_ulps_eq!(u[1], 0.);
    Ok(())
}
#[test]
fn trace_ray_multi() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let mut times = vec![];

    for i in 0..100 {
        let start = Instant::now();
        let miss_and_closesthit_ptx = r##"
.version 8.0
.target sm_86
.address_size 64

.entry __miss__ms() {
	.reg .b32 %r<6>;
	mov.b32 %r0, 0;
	mov.b32 %r1, 0;
        
	call _optix_set_payload, (%r0, %r1);
	ret;
}

.entry __closesthit__ch() {
	.reg .b32 %i<5>;
	.reg .b32 %v<5>;
	mov.b32 %i0, 0;
	mov.b32 %i1, 1;
	mov.b32 %i2, 2;
	mov.b32 %i3, 3;
	mov.b32 %i4, 4;

        mov.b32 %v0, 1;
	call _optix_set_payload, (%i0, %v0);
        
        call (%v1), _optix_read_primitive_idx, ();
	call _optix_set_payload, (%i1, %v1);
        
        call (%v2), _optix_read_instance_id, ();
	call _optix_set_payload, (%i2, %v2);
        
	.reg .f32 %f<2>;
        call (%f0, %f1), _optix_get_triangle_barycentrics, ();
        mov.b32 %v3, %f0;
        mov.b32 %v4, %f1;
	call _optix_set_payload, (%i3, %v3);
	call _optix_set_payload, (%i4, %v4);
        
	ret;
}
"##;
        let indices = ir.array(&[0u32, 1, 2])?;
        let vertices = ir.array(&[1.0f32, 0., 1., 0., 1., 1., 1., 1., 1.])?;

        let desc = AccelDesc {
            sbt: crate::SBTDesc {
                hit_groups: &[HitGroupDesc {
                    closest_hit: crate::ModuleDesc {
                        asm: miss_and_closesthit_ptx,
                        entry_point: "__closesthit__ch",
                    },
                    ..Default::default()
                }],
                miss_groups: &[MissGroupDesc {
                    miss: crate::ModuleDesc {
                        asm: miss_and_closesthit_ptx,
                        entry_point: "__miss__ms",
                    },
                }],
            },
            geometries: &[GeometryDesc::Triangles {
                vertices: &vertices,
                indices: &indices,
            }],
            instances: &[InstanceDesc {
                hit_group: 0,
                geometry: 0,
                transform: [1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0.],
            }],
        };
        let accel = ir.accel(desc)?;

        let payload = accel.trace_ray(
            &[
                &ir.literal(0u32)?,
                &ir.literal(0u32)?,
                &ir.literal(0u32)?,
                &ir.literal(0u32)?,
                &ir.literal(0u32)?,
            ],
            [
                &ir.array(&[0.6f32, 0.6])?,
                &ir.literal(0.6f32)?,
                &ir.literal(0.0f32)?,
            ],
            [
                &ir.literal(0.0f32)?,
                &ir.literal(0.0f32)?,
                &ir.array(&[1.0f32, -1.])?,
            ],
            &ir.literal(0.001f32)?,
            &ir.literal(1000.0f32)?,
            &ir.literal(0.0f32)?,
            None,
            None,
            None,
            None,
            None,
            None,
        )?;

        let valid = payload[0].cast(&VarType::Bool)?;

        let u = payload[3].bitcast(&VarType::F32)?;
        let v = payload[4].bitcast(&VarType::F32)?;

        valid.schedule();
        v.schedule();
        u.schedule();

        ir.eval()?;

        times.push(start.elapsed());
    }

    Ok(())
}
#[test]
fn trace_ray_hit_group() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;
    // pretty_env_logger::init();

    let miss_and_closesthit_ptx = r##"
.version 8.0
.target sm_86
.address_size 64

.entry __miss__ms() {
	.reg .b32 %r<6>;
	mov.b32 %r0, 0;
	mov.b32 %r1, 0;
        
	call _optix_set_payload, (%r0, %r1);
	ret;
}

.entry __closesthit__ch1() {
	.reg .b32 %i<5>;
	.reg .b32 %v<5>;
	mov.b32 %i0, 0;
	mov.b32 %i1, 1;
	mov.b32 %i2, 2;
	mov.b32 %i3, 3;
	mov.b32 %i4, 4;

        mov.b32 %v0, 1;
	call _optix_set_payload, (%i0, %v0);
        
	ret;
}

.entry __closesthit__ch2() {
	.reg .b32 %i<5>;
	.reg .b32 %v<5>;
	mov.b32 %i0, 0;
	mov.b32 %i1, 1;
	mov.b32 %i2, 2;
	mov.b32 %i3, 3;
	mov.b32 %i4, 4;

        mov.b32 %v0, 2;
	call _optix_set_payload, (%i0, %v0);
        
	ret;
}
"##;
    let indices0 = ir.array(&[0u32, 1, 2])?;
    let vertices0 = ir.array(&[1.0f32, 0., 1., 0., 1., 1., 0., 0., 1.])?;
    let indices1 = ir.array(&[0u32, 1, 2])?;
    let vertices1 = ir.array(&[1.0f32, 0., 1., 0., 1., 1., 1., 1., 1.])?;

    let desc = AccelDesc {
        sbt: crate::SBTDesc {
            hit_groups: &[
                HitGroupDesc {
                    closest_hit: crate::ModuleDesc {
                        asm: miss_and_closesthit_ptx,
                        entry_point: "__closesthit__ch1",
                    },
                    ..Default::default()
                },
                HitGroupDesc {
                    closest_hit: crate::ModuleDesc {
                        asm: miss_and_closesthit_ptx,
                        entry_point: "__closesthit__ch2",
                    },
                    ..Default::default()
                },
            ],
            miss_groups: &[MissGroupDesc {
                miss: crate::ModuleDesc {
                    asm: miss_and_closesthit_ptx,
                    entry_point: "__miss__ms",
                },
            }],
        },
        geometries: &[
            GeometryDesc::Triangles {
                vertices: &vertices0,
                indices: &indices0,
            },
            GeometryDesc::Triangles {
                vertices: &vertices1,
                indices: &indices1,
            },
        ],
        instances: &[
            InstanceDesc {
                hit_group: 0,
                geometry: 0,
                transform: [1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0.],
            },
            InstanceDesc {
                hit_group: 1,
                geometry: 1,
                transform: [1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0.],
            },
        ],
    };
    let accel = ir.accel(desc)?;

    let payload = accel.trace_ray(
        &[&ir.literal(0u32)?],
        [
            // &ir.array(&[0.5f32, 0.5])?,
            &ir.literal(0.5f32)?,
            &ir.literal(0.5f32)?,
            &ir.literal(0.0f32)?,
        ],
        [
            &ir.array(&[-0.2f32, 0.2])?,
            &ir.array(&[-0.2f32, 0.2])?,
            &ir.literal(1f32)?,
            // &ir.array(&[1.0f32, 1.])?,
        ],
        &ir.literal(0.001f32)?,
        &ir.literal(1000.0f32)?,
        &ir.literal(0.0f32)?,
        None,
        None,
        None,
        None,
        None,
        None,
    )?;

    let hit_id = payload[0].cast(&VarType::U32)?;
    hit_id.schedule();

    ir.eval()?;

    assert_eq!(hit_id.to_host::<u32>()?, [1, 2]);

    Ok(())
}
#[test]
fn trace_ray_shadowed() -> Result<()> {
    pretty_env_logger::try_init().ok();
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let miss_and_closesthit_ptx = r##"
.version 8.0
.target sm_86
.address_size 64

.entry __miss__ms() {
	.reg .b32 %r<6>;
	mov.b32 %r0, 0;
	mov.b32 %r1, 0;
        
	call _optix_set_payload, (%r0, %r1);
	ret;
}
.entry __miss__shadow() {
	.reg .b32 %r<6>;
	mov.b32 %r0, 0;
	mov.b32 %r1, 0;
        
	call _optix_set_payload, (%r0, %r1);
	ret;
}

.entry __closesthit__ch() {
	.reg .b32 %i<5>;
	.reg .b32 %v<5>;
	mov.b32 %i0, 0;
	mov.b32 %i1, 1;
	mov.b32 %i2, 2;
	mov.b32 %i3, 3;
	mov.b32 %i4, 4;

        mov.b32 %v0, 1;
	call _optix_set_payload, (%i0, %v0);
        
	ret;
}

"##;
    let indices0 = ir.array(&[0u32, 1, 2])?;
    let vertices0 = ir.array(&[1.0f32, 0., 1., 0., 1., 1., 0., 0., 1.])?;

    let desc = AccelDesc {
        sbt: crate::SBTDesc {
            hit_groups: &[HitGroupDesc {
                closest_hit: crate::ModuleDesc {
                    asm: miss_and_closesthit_ptx,
                    entry_point: "__closesthit__ch",
                },
                ..Default::default()
            }],
            miss_groups: &[
                MissGroupDesc {
                    miss: crate::ModuleDesc {
                        asm: miss_and_closesthit_ptx,
                        entry_point: "__miss__ms",
                    },
                },
                MissGroupDesc {
                    miss: crate::ModuleDesc {
                        asm: miss_and_closesthit_ptx,
                        entry_point: "__miss__shadow",
                    },
                },
            ],
        },
        geometries: &[GeometryDesc::Triangles {
            vertices: &vertices0,
            indices: &indices0,
        }],
        instances: &[InstanceDesc {
            hit_group: 0,
            geometry: 0,
            transform: [1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0.],
        }],
    };
    let accel = ir.accel(desc)?;

    let payload = accel.trace_ray(
        &[&ir.literal(0u32)?],
        [
            // &ir.array(&[0.5f32, 0.5])?,
            &ir.literal(0.5f32)?,
            &ir.literal(0.5f32)?,
            &ir.literal(0.0f32)?,
        ],
        [
            &ir.array(&[-0.2f32, 0.2])?,
            &ir.array(&[-0.2f32, 0.2])?,
            &ir.literal(1f32)?,
            // &ir.array(&[1.0f32, 1.])?,
        ],
        &ir.literal(0.001f32)?,
        &ir.literal(1000.0f32)?,
        &ir.literal(0.0f32)?,
        None,
        Some(&ir.literal(4u32)?),
        None,
        None,
        Some(&ir.literal(1u32)?),
        None,
    )?;

    let shadowed = payload[0].cast(&VarType::U32)?;
    shadowed.schedule();

    ir.eval()?;

    assert_eq!(shadowed.to_host::<u32>()?, [1, 0]);

    Ok(())
}
#[test]
fn sized_literal() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    let x = ir.sized_literal(0f32, 10)?;
    let x = x.add(&ir.literal(0f32)?)?;
    x.schedule();

    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(x.to_host::<f32>().unwrap(), vec![0f32; 10]);
    Ok(())
}
