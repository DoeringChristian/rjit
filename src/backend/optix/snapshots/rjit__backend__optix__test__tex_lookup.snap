---
source: src/backend/optix/test.rs
expression: ir.kernel_history()
---
Kernel History:
Launched CUDA Kernel 10287110750679701470 with 1 elements
===============================================
Kernel 10287110750679701470:

.version 8.0
.target sm_86
.address_size 64

.entry cujit(
	.param .align 8 .b8 params[64]) {

	.reg.b8   %b <21>; .reg.b16 %w<21>; .reg.b32 %r<21>;
	.reg.b64  %rd<21>; .reg.f32 %f<21>; .reg.f64 %d<21>;
	.reg.pred %p <21>;

	mov.u32 %r0, %ctaid.x;
	mov.u32 %r1, %ntid.x;
	mov.u32 %r2, %tid.x;
	mad.lo.u32 %r0, %r0, %r1, %r2; // r0 <- Index

	// Index Conditional (jump to done if Index >= Size).
	ld.param.u32 %r2, [params]; // r2 <- params[0] (Size)
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r0 >= r2
	@%p0 bra done; // if p0 => done
	
	mov.u32 %r3, %nctaid.x; // r3 <- nctaid.x
	mul.lo.u32 %r1, %r3, %r1; // r1 <- r3 * r1
	
body: // sm_86

	// [0]: ScheduleVar { op: Data, deps: [], ty: Void, reg: 5, data: Texture(0), sbt_hash: 0 } =>

	// [1]: ScheduleVar { op: Data, deps: [], ty: F32, reg: 7, data: Buffer(0), sbt_hash: 0 } =>

	// [2]: ScheduleVar { op: Literal, deps: [], ty: U32, reg: 8, data: Literal(0), sbt_hash: 0 } =>
	mov.b32 %r8, 0x0;


	// [3]: ScheduleVar { op: Literal, deps: [], ty: Bool, reg: 9, data: Literal(1), sbt_hash: 0 } =>
	mov.pred %p9, 0x1;


	// [4]: ScheduleVar { op: Gather, deps: [SVarId(1), SVarId(2), SVarId(3)], ty: F32, reg: 10, data: Buffer(1), sbt_hash: 0 } =>
	ld.param.u64 %rd0, [params+8];
	mad.wide.u32 %rd3, %r8, 4, %rd0;
	ld.global.nc.f32 %f10, [%rd3];

	// [5]: ScheduleVar { op: Data, deps: [], ty: F32, reg: 12, data: Buffer(2), sbt_hash: 0 } =>

	// [6]: ScheduleVar { op: Literal, deps: [], ty: U32, reg: 13, data: Literal(0), sbt_hash: 0 } =>
	mov.b32 %r13, 0x0;


	// [7]: ScheduleVar { op: Literal, deps: [], ty: Bool, reg: 14, data: Literal(1), sbt_hash: 0 } =>
	mov.pred %p14, 0x1;


	// [8]: ScheduleVar { op: Gather, deps: [SVarId(5), SVarId(6), SVarId(7)], ty: F32, reg: 15, data: Buffer(3), sbt_hash: 0 } =>
	ld.param.u64 %rd0, [params+24];
	mad.wide.u32 %rd3, %r13, 4, %rd0;
	ld.global.nc.f32 %f15, [%rd3];

	// [9]: ScheduleVar { op: TexLookup { dim: 2 }, deps: [SVarId(0), SVarId(4), SVarId(8)], ty: F32, reg: 16, data: None, sbt_hash: 0 } =>
	ld.param.u64 %rd0, [params+48];
	.reg.f32 %f16_out_<4>;
	tex.2d.v4.f32.f32 {%f16_out_0, %f16_out_1, %f16_out_2,
                             %f16_out_3}, [%rd0, {%f10, %f15}];

	// [10]: ScheduleVar { op: Extract { offset: 0 }, deps: [SVarId(9)], ty: F32, reg: 17, data: Buffer(4), sbt_hash: 0 } =>
	mov.b32 %f17, %f16_out_0;

	// [11]: ScheduleVar { op: Idx, deps: [], ty: U32, reg: 18, data: None, sbt_hash: 0 } =>
	mov.u32 %r18, %r0;


	// [12]: ScheduleVar { op: Literal, deps: [], ty: Bool, reg: 19, data: Literal(1), sbt_hash: 0 } =>
	mov.pred %p19, 0x1;


	// [13]: ScheduleVar { op: Scatter { op: None }, deps: [SVarId(10), SVarId(10), SVarId(11), SVarId(12)], ty: F32, reg: 20, data: None, sbt_hash: 0 } =>
	ld.param.u64 %rd0, [params+40];
	mad.wide.u32 %rd3, %r18, 4, %rd0;
	st.global.f32 [%rd3], %f17;

	//End of Kernel:

	add.u32 %r0, %r0, %r1; // r0 <- r0 + r1
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r1 >= r2
	@!%p0 bra body; // if p0 => body


done:

	ret;
}

